wgpu-core/src/command/render.rs

```diff
@@ -5,23 +5,165 @@
     command::{
         bind::{Binder, LayoutChange},
-        CommandBuffer,
+        PhantomSlice,
+        RawRenderTargets,
     },
     conv,
-    device::{RenderPassContext, BIND_BUFFER_ALIGNMENT, MAX_VERTEX_BUFFERS},
-    hub::{GfxBackend, Global, IdentityFilter, Token},
-    id::{BindGroupId, BufferId, CommandBufferId, RenderPassId, RenderPipelineId},
+    device::{
+        FramebufferKey,
+        RenderPassContext,
+        RenderPassKey,
+        BIND_BUFFER_ALIGNMENT,
+        MAX_VERTEX_BUFFERS,
+        MAX_COLOR_TARGETS,
+    },
+    hub::{GfxBackend, Global, Token},
+    id,
     pipeline::{IndexFormat, InputStepMode, PipelineFlags},
-    resource::BufferUsage,
+    resource::{BufferUsage, TextureUsage, TextureViewInner},
     track::TrackerSet,
     BufferAddress,
     Color,
     Stored,
 };
 
+use arrayvec::ArrayVec;
 use hal::command::CommandBuffer as _;
+use peek_poke::{Peek, PeekCopy, Poke};
+
+use std::{
+    borrow::Borrow,
+    collections::hash_map::Entry,
+    iter,
+    marker::PhantomData,
+    ops::Range,
+};
+
 
-use std::{iter, ops::Range};
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub enum LoadOp {
+    Clear = 0,
+    Load = 1,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub enum StoreOp {
+    Clear = 0,
+    Store = 1,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct RenderPassColorAttachmentDescriptorBase<T, R> {
+    pub attachment: T,
+    pub resolve_target: R,
+    pub load_op: LoadOp,
+    pub store_op: StoreOp,
+    pub clear_color: Color,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug)]
+pub struct RenderPassDepthStencilAttachmentDescriptorBase<T> {
+    pub attachment: T,
+    pub depth_load_op: LoadOp,
+    pub depth_store_op: StoreOp,
+    pub clear_depth: f32,
+    pub stencil_load_op: LoadOp,
+    pub stencil_store_op: StoreOp,
+    pub clear_stencil: u32,
+}
+
+//Note: this could look better if `cbindgen` wasn't confused by &T used in place of
+// a generic paramter, it's not able to mange
+pub type OptionRef<'a, T> = Option<&'a T>;
+pub type RenderPassColorAttachmentDescriptor<'a> =
+    RenderPassColorAttachmentDescriptorBase<id::TextureViewId, OptionRef<'a, id::TextureViewId>>;
+pub type RenderPassDepthStencilAttachmentDescriptor =
+    RenderPassDepthStencilAttachmentDescriptorBase<id::TextureViewId>;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct RenderPassDescriptor<'a> {
+    pub color_attachments: *const RenderPassColorAttachmentDescriptor<'a>,
+    pub color_attachments_length: usize,
+    pub depth_stencil_attachment: Option<&'a RenderPassDepthStencilAttachmentDescriptor>,
+}
+
+#[derive(Clone, Copy, Debug, PeekCopy, Poke)]
+pub struct Rect<T> {
+    pub x: T,
+    pub y: T,
+    pub w: T,
+    pub h: T,
+}
+
+#[derive(Clone, Copy, Debug, PeekCopy, Poke)]
+enum RenderCommand {
+    SetBindGroup {
+        index: u8,
+        num_dynamic_offsets: u8,
+        bind_group_id: id::BindGroupId,
+        phantom_offsets: PhantomSlice<BufferAddress>,
```

Comment:
I found it a bit confusing to add `PhantomSlice` to the end of these variants because I wasn't sure if they were markers for `peek_poke` to know to treat them specially somehow.

Instead of using this marker type, could we replace it by comments to indicate that there is some dynamically sized array at the end of these variants and have `decode` as a free function?

Reply:
I think having at least some type-level semantic of the data following commands is helpful. having the `decode` strictly associated with a `PhantomSlice` makes it harder to fail when writing and maintaining this code. Comments wouldn't be as helpful.

Reply:
I guess one interesting (future) option would be for peek_poke to provide a marker type to declare the size of variable length arrays, i.e.
```rust
#[derive(PeekCopy, Poke)]
enum RenderCommand {
    SetBindGroup {
        index: u8,
        num_dynamic_offsets: u8,
        #[peek_poke(array_size = "num_dynamic_offsets")]
        dynamic_offsets: PeekPoke::VariableLengthArray<BufferAddress>,
    },
    /* ... */
}
```